zerver/lib/actions.py

```diff
@@ -5532,17 +5532,24 @@ def do_update_message_flags(
         if flag != "starred":
             raise JsonableError(_("Invalid message(s)"))
         # Validate that the user could have read the relevant message
-        message = access_message(user_profile, messages[0])[0]
-
-        # OK, this is a message that you legitimately have access
-        # to via narrowing to the stream it is on, even though you
-        # didn't actually receive it.  So we create a historical,
-        # read UserMessage message row for you to star.
-        UserMessage.objects.create(
-            user_profile=user_profile,
-            message=message,
-            flags=UserMessage.flags.historical | UserMessage.flags.read,
-        )
+        with transaction.atomic():
```

Comment:
Could the `transaction.atomic()` be around a larger block, such that we don't need to do a second UserMessage check?

Reply:
I don't think so. The problem really is that we lock the message *after* we realize there's no UserMessage present.

Reply:
I adjusted the code to avoid the new query -- `access_message` already queries for `UserMessage` anyways 😅